Add support for API streaming to the watch-cache

p0lyn0mial opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
This PR implements API streaming for the watch cache as described in https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3157-watch-list#design-details

xref: https://github.com/kubernetes/enhancements/issues/3157
xref: https://github.com/kubernetes/enhancements/pull/3142
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Implements API for streaming for the watch-cache

When sendInitialEvents ListOption is set together with watch=true, it begins the watch stream with synthetic init events followed by a synthetic "Bookmark" after which the server continues streaming events.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3157-watch-list
```

p0lyn0mial commented:
/assign @wojtek-t 
/cc @sttts

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:
```diff
@@ -17,17 +17,19 @@ limitations under the License.
 
 // ValidateListOptions returns all validation errors found while validating the ListOptions.
-func ValidateListOptions(options *internalversion.ListOptions) field.ErrorList {
+func ValidateListOptions(options *internalversion.ListOptions, isWatchListFeatureEnabled bool) field.ErrorList {
```
I don't understand why you're passing this parameter explicitly.

You should just use the value of the feature gate directly below (and if you need to test it, just switch the feature gate appropriate in your tests, e.g.
```
defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.WatchList, true)()
```

p0lyn0mial replied:
I did it to avoid importing `k8s.io/apiserver`. I think that `apimachinery` pkg should not depend on the `apiserver` pkgs

wojtek-t replied:
This has been discussed in API PR already.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:
```diff
@@ -44,3 +46,23 @@ func ValidateListOptions(options *internalversion.ListOptions) field.ErrorList {
+		if match != metav1.ResourceVersionMatchMostRecent {
+			allErrs = append(allErrs, field.NotSupported(field.NewPath("resourceVersionMatch"), match, []string{string(metav1.ResourceVersionMatchMostRecent)}))
+		}
+		if match == metav1.ResourceVersionMatchMostRecent && !options.AllowWatchBookmarks {
```
Shouldn't we force watch bookmarks even if the "MostRecent" param is not set?

wojtek-t replied:
ping

p0lyn0mial replied:
I don't think so. Why we should force clients to register for progress notification? 

Here we say that if you request `MostRecent` then you also need to understand bookmarks events.
I think we should also check if the feature flag for bookmarks was enabled.

p0lyn0mial replied:
oh, `WatchBookmark` is `LockToDefault` so we are good.

wojtek-t replied:
OK - I initially thought that we may want to support "list via watch" also for "NotOlderThan".
If we limit ourselves to MostRecent - then it's fine.

But I'm still chewing if we don't want to support the "NotOlderThan" for  "list via watch".


In other words - I'm thinking (I'm far from being convinced at this point, but I'm thinking) about extending ListOptions with the new field:
  `SendInitialEvents bool`

Default to false would give us backward compatibility.
If explicitly set to true (for watch requests) it would basically be sending the initial watch events according to the existing semantic of ResourceVersionMatch:
- NotOlderThan (you get bookmark when we pass the given RV)
- Exact (you get bookmark when we pass then "current etcd RV")

wojtek-t replied:
You may want to ask @deads2k for his opinion which API he would prefer, I'm not an API expert.

p0lyn0mial replied:
for the LIST we could also use `MostRecent` in that case we would wait until the cache has caught up and only then send initial data (without the bookmark event). I think that would fix the long-standing issue with consistency

wojtek-t replied:
i think we're mixing two things here with this API.

I would like to allow passing the RV myself too. In other words, I would like to be able to say that instead of using MostRecent, I would like to use watch-based list but pass that RV myself (e.g. explicitly pass 100).

So I still think that an explicitly `SendInitialEvents bool` field might be a better option.

p0lyn0mial replied:
To summarise our discussion:

We have to use `resourceVersionMatch`. 

Passing a new param (i.e. '/api/v1/namespaces/kube-system/pods?abc=xyz&watch=true') doesn't ensure backward compatibility. We won't get a 422 response. While we could patch prior versions of k8s doing so for aggregated APIs is impossible.

Given the above we have at least two options:

Option 1: Define a new resourceVersionMatch option (`MostRecent`) for WATCH requests. Allow to use it with **any** RV value.

Option 2: Define a new resourceVersionMatch option (`MostRecent`) for WATCH requests. Allow to use it **only** with RV=0 | RV="". This would allow to use it only for initial requests.


My preference is towards the `Option 1`. However, before we merge this PR I'd like to gather some feedback.

p0lyn0mial replied:
@deads2k what is your opinion/preference?

deads2k replied:
> OK - I initially thought that we may want to support "list via watch" also for "NotOlderThan".

I agree with this sentiment: I expected this to work as well.  `SendInitialEvents` seems to more cleanly describe what we want to happen.  That would allow the following combinations (correct me if I'm mistaken)

1. sendInitialEvent and rv>0 - this would provide the initial state and when we reach the desired RV, the synthetic bookmark (this must be recognizably different) will let you know when you are done with the list.
2. sendInitialEvent and rv=0 - when the initial state has been replayed as far as we have available, the synthetic bookmark will let you know when you are done with the list.
3. sendInitialEvent and rv="" - when the initial state has been replayed as far as etcd has progressed (MostRecent), the synthetic bookmark will let you know when you are done with the list.

It appears to me that the new ResourceVersionMatch value is being used here to allow old servers to indicate to clients that they do not support the option the client is requesting.  We have historically managed this by explicitly rejecting new options in older version.  I remember a PR doing that same thing for `DryRun`.  The choice of adding a ResourceVersionMatch is more immediately convenient, but less practical for numerous features over time as the temptation to continue overloading enums or existing options grows.  See https://github.com/kubernetes/kubernetes/pull/63557 for an example.

p0lyn0mial replied:
> 1. sendInitialEvent and rv>0 - this would provide the initial state and when we reach the desired RV, the synthetic bookmark (this must be recognizably different) will let you know when you are done with the list.

why we would like to provide the initial state in this case? I think in that case we should only give you events that happened after the provided RV. I think this would scale better than replying the initial events the client has already seen.

> It appears to me that the new ResourceVersionMatch value is being used here to allow old servers to indicate to clients that they do not support the option the client is requesting

yes, we don't want to provide any compatibilities for the aggregated APIs?

wojtek-t replied:
Thanks @deads2k for a nice summary - it exactly reflects my thinking (although I wasn't sure what's the recommended path for doing that from the API perspective).

> why we would like to provide the initial state in this case? I think in that case we should only give you events that happened after the provided RV. I think this would scale better than replying the initial events the client has already seen.

There are two cases:
- you specify the new `SendInitialEvents` parameter - then I want to see initial events. The usecase for that is that e.g. my reflector went out of history window so I can't resume my watch and need to relist. But the only think I need to ensure is to not go back in time, so I'm fine with slightly delayed data - i.e. NotOlderThan with rv=X
- you don't specify the new `SendInitialEvents` parameter - then I just don't want to see the initial events.

So as I originally pointed out - I'm actually with David wrt it.

> yes, we don't want to provide any compatibilities for the aggregated APIs?

Thinking about this more - maybe it's not strictly needed immediately? We're generally saying that you should first upgrade your apiservers and only then upgrade clients.
What if we would just:
- introduce the new option now (and reject the request if the field is not supported (based on feature gate)
- not turn on this feature in any of our libraries by default for the next N releases (you can only enable it explicitly if you know what you're doing, e.g. if you're using built-in APIs).
Would that be an acceptable guard?

p0lyn0mial replied:
okay, I've updated the PR to reflect the above, please check if this is what we want.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -418,6 +418,12 @@ const (
+	// For watch calls, it begins with synthetic "Added" events of all resources up to the most recent ResourceVersion.
+	// It ends with a synthetic "Bookmark" event containing the most recent ResourceVersion.
+	// For list calls, it has the same semantics as leaving ResourceVersion and ResourceVersionMatch unset.
+	ResourceVersionMatchMostRecent ResourceVersionMatch = "MostRecent"
```
Note to myself: think about it a bit more how to define the semantic the best here (and where this is the right option).

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -293,9 +377,11 @@ type Cacher struct {
 	// Maintain a timeout queue to send the bookmark event before the watcher times out.
 	bookmarkWatchers *watcherBookmarkTimeBuckets
+
+	bookmarkAfterRVWatchers *bookmarkAfterResourceVersionWatchers
```
Please correct me if I'm missing something, but I'm not sure I like the current approach.
This is introducing a bunch of new code and fields and I'm not convinced we really need that (vs putting this functionality elsewhere).

1. We need a way to distinguish a watch that should be running "until a given RV" from the regular ones.
I would much rather prefer to put that login into `cacheWatcher` directly and let it handle everything inside.
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L1148
[cacheWatcher is much simpler thing to follow then the overall Cacher and it would simplify the reasoning, debugging, etc.]

2. The way I'm currently thinking about it is:
- cacheWatcher gets an additional parameter of "endRV' (better name needed)
- the overall Cacher remains unchanged
- inside cacheWatcher we add the logic of proper handling termination, so at the high level (I'm sure things are missing here)
  (a) when it gets bookmark and isn't yet synced, just ignore it
  (b) if it gets event with RV higher than its EndRV, just ignore it a stop itself

[With the above we're technically risking having some already stopped watchers in bookmarkWatchers, but that's fine - this is possible even now - they are just GCed when they bookmark time approaches.]

Please take a look and let me know what do you think about it.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -916,8 +938,7 @@ func (c *Cacher) startDispatchingBookmarkEvents() {
 			c.watchersBuffer = append(c.watchersBuffer, watcher)
-			// Requeue the watcher for the next bookmark if needed.
-			c.bookmarkWatchers.addWatcher(watcher)
+			c.expiredBookmarkWatchers = append(c.expiredBookmarkWatchers, watcher)
```
Trying to minimize the core PR again. This is effectively a refactoring that we can do independently.
Can you maybe move the no-op changes to a separate PR:
- the introduction of expiredBookmarkWatchers
- rename of function XxxLocked
- most of comment updated in: nextBookmarkTime [eventually let's add your new condition as (c) and let's make that incremental over the no-op PR factored out of this one]
-

p0lyn0mial replied:
sure, np

p0lyn0mial replied:
done https://github.com/kubernetes/kubernetes/pull/111244

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1222,7 +1302,9 @@ func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer) bool {
-		c.forget()
+
+		// TODO: determine if the watcher needs draining
+		c.forget(false)
```
I started wondering if we will really need that option at all or whether the watcher will be able to determine it themselves.

I think this all boils down the answering the question when exactly the watcher should be draining.
The only situation when I believe it makes sense is basically the situation when:
- the watcher is requesting the "bookmark after sync"
- that bookmark is already in its channel

In other situations I think it doesn't bring any value. Does that make sense?
If so - then the watcher can effectively figure it out themselves and we won't be forced to bass any arguments anywhere.

WDYT?

p0lyn0mial replied:
I think it is necessary, mainly because we need to cover the following cases:

1. client disconnect (watcher.Stop())
2. c.terminateAllWatchers()

So, even if we are in a draining mode, triggering any of the above conditions should stop the watcher.
Does it make sense?

wojtek-t replied:
OK - I think it makes sense.
So I think you should now figure out the condition here and document it clearly in a comment.

p0lyn0mial replied:
yes, ok.

wojtek-t replied:
This now looks reasonable to me - resolving.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:
```diff
@@ -17,17 +17,19 @@ limitations under the License.
 package validation
 
 import (
+	"fmt"
```
nit: we should probably structure the PR commits into:
- feature gate
- API change (together with validation)
- changes to watchcache

wojtek-t reviewed (commented):
Didn't yet look into tests themselves, but the rest looks reasonable to me - added some comments.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1068,6 +1088,41 @@ func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
+		Field: fields.Everything(),
+		Limit: 1, // just in case we actually hit something
+	}
+	err := c.storage.GetList(ctx, "/reserved_non_existing_prefix", storage.ListOptions{Predicate: pred}, emptyList)
```
Hmm - I would actually go with a real key rather then the artifical one.

wojtek-t replied:
As discussed offline - I don't feel very strongly about that.
Let's keep it as is and let's add a TODO, to reevaluate it before going to Beta.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1176,21 +1231,38 @@ type cacheWatcher struct {
+		objectType:                   objectType,
+		identifier:                   identifier,
+		bookmarkAfterResourceVersion: bookmarkAfterResourceVersion,
+		wasBookmarkAfterResourceVersionSent: func() bool {
```
wasBookmarkAfterResourceVersionSend: (bookmarkAfteResourceVersion == 0),

:)

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1201,21 +1273,32 @@ func (c *cacheWatcher) ResultChan() <-chan watch.Event {
 }
 
 func (c *cacheWatcher) nonblockingAdd(event *watchCacheEvent) bool {
+	// if the bookmarkAfterResourceVersion hasn't been seen we will try to deliver a bookmark event every second.
```
nit: can you make those comment lines shorter?

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1231,11 +1314,30 @@ func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer) bool {
-		// we simply terminate it.
+		// Since we don't want to block on it infinitely, we simply terminate it.
+
+		// if the bookmarkAfterResourceVersionReceived hasn't been received we can safely terminate the watcher.
```
nit: If ... (start with a capital letter).
Also please make the comment lines shorter

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -105,17 +106,17 @@ func (wm watchersMap) addWatcher(w *cacheWatcher, number int) {
 }
 
-func (wm watchersMap) deleteWatcher(number int, done func(*cacheWatcher)) {
+func (wm watchersMap) deleteWatcher(number int, drainWatcher bool, done func(bool, *cacheWatcher)) {
```
OK - I think I finally figured out why I didn't like this.
Basically, the watchersMap doesn't use the information whether the watch should be drained or not (other than passing it to the watch itself).

So they way I would like this work is to:
- revert the changes to all watchersMap (those should just be taking func(*cacheWatcher) as before
- add a field to cacheWatcher like `shouldDrain`
- switch the `forget()` function to basically be sth like:
```
func forget(bool) {
  watcher.setShouldDrain(bool)
  deleteWatcher(...) // as called now without drain parameter
}
```

- and then the cacheWatcher::stopLocked() will be looking at the `shouldDrain` param and choosing what to do based on that

wojtek-t reviewed (commented):
Couple more comments - but I think we're getting there

Please restructure the commits so that we can actually show it to David for his opinion.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1176,21 +1228,42 @@ type cacheWatcher struct {
+	// and send all event in the input buffer.
+	drainInputBuffer bool
+
+	// bookmarkAfterResourceVersion holds an RV that indicates when we should start delivering bookmark events
```
Brainstorming:

We effectively have a state machine here with states:
- WAITING_FOR_BOOKMARK
- BOOKMARK_RECEIVED
- BOOKMARK_SENT

I'm wondering if storing that wouldn't be more intuitive:
- we have a field "bookmarkAfterResourceVersion" (which is effectively immutable)
- we have a state (that is mutable):
  (a) in waiting state and got high-enough bookmark, we transition to boomark-received
  (b) if in bookmar-receivced state and sending bookmark transition to bookmar-sent state
  (c) bookmar-send is a terminal state

wojtek-t replied:
Thanks for doing this!

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1201,21 +1274,32 @@ func (c *cacheWatcher) ResultChan() <-chan watch.Event {
+	}
 	select {
 	case c.input <- event:
+		c.markBookmarkAfterResourceVersionAsReceived(event)
```
The same thing is blocking in the c.input <- call in the `add` method.

p0lyn0mial replied:
no, bookmarks events are only added via `nonblockingAdd` method

wojtek-t replied:
OK - good point. Can you please add a comment for add() method that Bookmark events are never using add?

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1445,7 +1614,9 @@ func (c *cacheWatcher) process(ctx context.Context, resourceVersion uint64) {
 			// only send events newer than resourceVersion
 			if event.ResourceVersion > resourceVersion {
-				c.sendWatchCacheEvent(event)
+				if c.sendWatchCacheEvent(event) {
```
instead of changing the signature of sendWatchCacheEvent, let's call the c.markBookmarkAfter... directly from that only place where true is returned.

p0lyn0mial replied:
we could do that.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1201,21 +1269,35 @@ func (c *cacheWatcher) ResultChan() <-chan watch.Event {
+	// if it is < than the bookmarkAfterResourceVersion
+	// so that we don't pollute the input channel
+	if event.Type == watch.Bookmark && !c.wasBookmarkAfterResourceVersionReceived() && event.ResourceVersion < c.bookmarkAfterResourceVersion {
+		return false // skip this bookmark event
```
nit: remove this "//skip ..." comment - it's obvious

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1231,11 +1313,32 @@ func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer) bool {
+		// as clients can re-establish a new watch with the given RV and receive
+		// further notifications.
+		if !c.wasBookmarkAfterResourceVersionSentLocked() {
+			c.forget(true)
```
I think the metric should be updated here too - in the end we're terminating this watcher too (due to its slowness), just doing that gracefully.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1231,11 +1313,32 @@ func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer) bool {
+		// further notifications.
+		if !c.wasBookmarkAfterResourceVersionSentLocked() {
+			c.forget(true)
+			return
```
Same about the log - we should also write it in all cases, just add an additional info whether this deletion is graceful or not.

So I would do something like (the more I'm thinking about it, the more I think that the state-machine idea is a good one).
```
// If we didn't even receive the bookmark, we can safely terminate the watcher.
// If we received the bookmark but didn't yet send it, we should try to send it by shutting
// down gracefully to allow the client to make the progress.
// If we already send a bookmark, we can again safely terminate the watcher.
graceful := c.state == BOOKMARK_RECEIVED
klog.V(1).Infof("Forcing ... graceful=%t ...)
c.foget(graceful)
```

wojtek-t replied:
With your current changes from the last commit - this would now simplify the code here.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1581,3 +1585,219 @@ func TestCacheIntervalInvalidationStopsWatch(t *testing.T) {
+
+func TestBookmarkAfterResourceVersionWatchers(t *testing.T) {
+	newWatcher := func(id string, deadline time.Time) *cacheWatcher {
+		return newCacheWatcher(0, func(_ string, _ labels.Set, _ fields.Set) bool { return true }, func(bool) {}, testVersioner{}, deadline, true, objectType, id, 10)
```
can you please define:
```
func everything(_ string, _ labels.Set, _ fields.Set) bool {
  return true
}
```

[optional a bonus use across the whole file]

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1581,3 +1585,219 @@ func TestCacheIntervalInvalidationStopsWatch(t *testing.T) {
+
+	clock := testingclock.NewFakeClock(time.Now())
+	target := newTimeBucketWatchers(clock, defaultBookmarkFrequency)
+	target.addWatcher(newWatcher("1", clock.Now().Add(2*time.Minute)))
```
check that it returns true

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1581,3 +1585,219 @@ func TestCacheIntervalInvalidationStopsWatch(t *testing.T) {
+	target := newTimeBucketWatchers(clock, defaultBookmarkFrequency)
+	target.addWatcher(newWatcher("1", clock.Now().Add(2*time.Minute)))
+
+	// the watcher is immediately expired
```
in this case, nextBookmarkTime returns `time.Time{}, true`

But then what happens we're doing:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L190

so we're effectively getting negative duration. This works but is a bit unexpected to me.
Maybe lets add a comment in the nextBookmarkTime method ardoun that line that we can actually get negative duration and that works fine and is expected.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1581,3 +1585,219 @@ func TestCacheIntervalInvalidationStopsWatch(t *testing.T) {
+
+	// the watcher is immediately expired
+	ret := target.popExpiredWatchers()
+	if len(ret) != 1 {
```
`|| len(ret[0]) != 1`

wojtek-t replied:
same in places below

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1581,3 +1585,219 @@ func TestCacheIntervalInvalidationStopsWatch(t *testing.T) {
+	if len(ret) != 1 {
+		t.Fatalf("expected only one watcher to be expired")
+	}
+	target.addWatcher(ret[0][0])
```
check the value?
same below

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1581,3 +1585,219 @@ func TestCacheIntervalInvalidationStopsWatch(t *testing.T) {
+	}
+	w = newCacheWatcher(1, filter, forget, testVersioner{}, time.Now(), true, objectType, "", 0)
+	go w.processInterval(context.Background(), intervalFromEvents(initEvents), 0)
+	if !w.add(&watchCacheEvent{Object: &v1.Pod{}}, time.NewTimer(1*time.Second)) {
```
hmm - how is this even treated? it doesn't set RV, type, anything like that...

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1581,3 +1585,219 @@ func TestCacheIntervalInvalidationStopsWatch(t *testing.T) {
+	}
+	if !w.add(&watchCacheEvent{Object: &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "p2"}}, ResourceVersion: 15}, time.NewTimer(1*time.Second)) {
+		t.Fatal("failed adding an even to the watcher")
+	}
```
Why 3 events can be added?
The channel size is 2 so:
- the 2 init events will fill up the resultCh
- the 2 further events will fill in inputCh
- where is the 3rd event?

wojtek-t replied:
ohh ok - it's actually kept in memory by process() hanging to put it into resultCh...

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1581,3 +1585,219 @@ func TestCacheIntervalInvalidationStopsWatch(t *testing.T) {
+		t.Fatalf("expected to get a bookmark event but got %v", e)
+	}
+	<-w.ResultChan() // w.add p2
+	<-w.ResultChan() // w.add p3
```
wait - p3 wasn't added above (it add would return true, Fatal() would be called..

So what do we get here?

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1212,6 +1212,18 @@ func (c *errWatcher) Stop() {
 
+// possible states of the cache watcher
+const (
+	// cacheWatcherWaitingForBookmark indicates the cacher is waiting for a bookmark event with a specific RV set
```
nit: please shorten the lines a bit

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1241,28 +1253,33 @@ type cacheWatcher struct {
 
-	// wasBookmarkAfterResourceVersionSentMutex protects wasBookmarkAfterResourceVersionSent
-	wasBookmarkAfterResourceVersionSentMutex sync.Mutex
+	// sateMutex protects state
```
s/sate/state/

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1241,28 +1253,33 @@ type cacheWatcher struct {
-	// wasBookmarkAfterResourceVersionSentMutex protects wasBookmarkAfterResourceVersionSent
-	wasBookmarkAfterResourceVersionSentMutex sync.Mutex
+	// sateMutex protects state
+	sateMutex sync.Mutex
```
We could consider using atomic instead too - but I don't have strong preference. Mutex+int works for me too.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1241,28 +1253,33 @@ type cacheWatcher struct {
+		objectType:                   objectType,
+		identifier:                   identifier,
+		bookmarkAfterResourceVersion: bookmarkAfterResourceVersion,
+		state: func() int {
```
nit: let's move it out of here?
```
state := cacheWatcherBookmarkSent
if bookmarkAfterResourceVersion != 0 {
  state = cacheWatcherWaitingForBookmark
}
return &cacheWatcher{
  ...
```

wojtek-t replied:
ping

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1230,12 +1333,34 @@ func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer) bool {
+		// If the bookmarkAfterResourceVersionReceived hasn't been received,
+		// we can safely terminate the watcher. Because the client is waiting
+		// for this specific bookmark, and we even haven't received one.
+		if !c.wasBookmarkAfterResourceVersionReceived() {
```
Can you please extract this logic to a separate dedicated function like:
```
func (c *cacheWatcher) shouldCloseGracefull() {
 ...
}
```

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1281,6 +1415,61 @@ func (c *cacheWatcher) nextBookmarkTime(now time.Time, bookmarkFrequency time.Du
 
+// wasBookmarkAfterResourceVersionReceived checks if the given cacheWatcher have seen a bookmark event >=  bookmarkAfterResourceVersion
+func (c *cacheWatcher) wasBookmarkAfterResourceVersionReceived() bool {
+	return c.state == cacheWatcherBookmarkReceived || c.state == cacheWatcherBookmarkSent
```
nit:

c.state != cacheWatcherWaitingForBookmark

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1281,6 +1415,61 @@ func (c *cacheWatcher) nextBookmarkTime(now time.Time, bookmarkFrequency time.Du
 }
 
+// wasBookmarkAfterResourceVersionReceived checks if the given cacheWatcher have seen a bookmark event >=  bookmarkAfterResourceVersion
+func (c *cacheWatcher) wasBookmarkAfterResourceVersionReceived() bool {
```
Either take a lock or call the function `...Locked`

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1281,6 +1415,61 @@ func (c *cacheWatcher) nextBookmarkTime(now time.Time, bookmarkFrequency time.Du
+		// since we received a bookmark event that means we have
+		// converged towards the expected RV and it is okay to update the state so that
+		// this cacher can be scheduler for a regular bookmark events
+		c.state = cacheWatcherBookmarkReceived
```
Not thread safe.

wojtek-t reviewed (commented):
I added a lot of comments, but I like the current path much more - it's much easier to follow imho (and can be simplified furrther as I pointed in my comments).

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1201,41 +1293,78 @@ func (c *cacheWatcher) ResultChan() <-chan watch.Event {
+		// (a) The bookmarkAfterResourceVersionReceived hasn't been received,
+		//     we can safely terminate the watcher. Because the client is waiting
+		//     for this specific bookmark, and we even haven't received one.
+
```
nit: remove this empty line

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1281,6 +1419,71 @@ func (c *cacheWatcher) nextBookmarkTime(now time.Time, bookmarkFrequency time.Du
 }
 
+// wasBookmarkAfterResourceVersionReceivedLocked same as wasBookmarkAfterResourceVersionReceived just acquires a lock
+func (c *cacheWatcher) wasBookmarkAfterResourceVersionReceivedLocked() bool {
```
Remove `Locked` - we use `Locked` in the name when the lock is already held when this function is being called.

This also means updating names of the functions below.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1176,21 +1247,42 @@ type cacheWatcher struct {
+
+	// bookmarkAfterResourceVersion holds an RV that indicates when we should start delivering bookmark events
+	// if this field holds the value of 0 that means we don't have any special preferences toward delivering bookmark events
+	bookmarkAfterResourceVersion uint64
```
Let's maybe state explicitly that this field is assumed to be immutable for the entire life of cacheWatcher.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -496,6 +505,16 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 
+	// Determine if a consistent stream was requested
+	bookmarkAfterResourceVersion := uint64(0)
+	if watchRV == 0 && opts.ResourceVersionMatch == metav1.ResourceVersionMatchMostRecent && utilfeature.DefaultFeatureGate.Enabled(features.WatchList) {
```
It's more a question about the API than the code here - but I would like to allow passing the RV myself too.
In other words, I would like to be able to say that instead of using MostRecent, I would like to use watch-based list but pass that RV myself (e.g. explicitly pass 100).
Will copy the comment to the API one discussion too.

p0lyn0mial replied:
pass an RV = 100 and still require `MostRecent` semantics (i.e. consistency) ?

wojtek-t replied:
No. Pass a "RV=100 and expect a bookmark after I synced to RV=100+ (maybe 100, maybe 101, maybe 110, etc. But I don't care about what the most recent version is).

Basically support "list via watch" for the current "NotOlderThan" semantic.

p0lyn0mial replied:
you can do that, in that case, you should not wait for a bookmark event, you will get one after ~60s.

wojtek-t replied:
But that's a huuuge regression comparing to current "list+watch" pattern. I wouldn't be fine with that as a client.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1176,21 +1247,43 @@ type cacheWatcher struct {
+	// and send all event in the input buffer.
+	drainInputBuffer bool
+
+	// bookmarkAfterResourceVersion holds an RV that indicates when we should start delivering bookmark events
```
nit: shorten the comment lines a bit

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1201,41 +1294,77 @@ func (c *cacheWatcher) ResultChan() <-chan watch.Event {
+
+		c.stateMutex.Lock()
+		defer c.stateMutex.Unlock()
+		graceful := c.wasBookmarkAfterResourceVersionReceivedLocked() && !c.wasBookmarkAfterResourceVersionSent()
```
Let's switch to:
```
graceful := (c.state == cacheWatcherBookmarkReceived)
```

Seems easier to follow and understand (especially given you take the lock explicitly here anyway.

wojtek-t replied:
Also switch to:
```
graceful := func() {
  c.stateMute.Lock()
  defer c.stateMutex.Unlock()
  return c.state == cacheWatcherBookmarkReceived
}()
```

[avoids locking for the time of calling forget]

p0lyn0mial replied:
we are graceful when we have received an RV but i hasn't been sent, so we cannot simply check `graceful := (c.state == cacheWatcherBookmarkReceived)`

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1201,41 +1294,77 @@ func (c *cacheWatcher) ResultChan() <-chan watch.Event {
+		//     That means we should drain the input buffer which contains
+		//     the bookmarkAfterResourceVersion we want. We do that to make progress
+		//     as clients can re-establish a new watch with the given RV and receive
+		//     further notifications.
```
Given my comment below let's maybe shorten this whole comment to something like:
```
We set graceful=true only if we've seen bookmark but haven't yet send that to the client.
Otherwise we can safely terminate the watcher.
```

[And remove the graceful = false part too]

wojtek-t replied:
ping (and in couple different places)

wojtek-t replied:
ping

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1281,6 +1419,71 @@ func (c *cacheWatcher) nextBookmarkTime(now time.Time, bookmarkFrequency time.Du
 	return heartbeatTime, true
 }
 
+// wasBookmarkAfterResourceVersionReceived same as wasBookmarkAfterResourceVersionReceivedLocked just acquires a lock
```
nit: shorten the comment lines here (and in below functions)

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1281,6 +1419,71 @@ func (c *cacheWatcher) nextBookmarkTime(now time.Time, bookmarkFrequency time.Du
+}
+
+// wasBookmarkAfterResourceVersionSent checks if a bookmark event with an RV >= the bookmarkAfterResourceVersion has been sent by this watcher
+func (c *cacheWatcher) wasBookmarkAfterResourceVersionSent() bool {
```
After applying my comments from above - I suggest inlining the only remaining call to this function (in markBookmarkAfter.. and getting rid of this function.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1201,41 +1296,78 @@ func (c *cacheWatcher) ResultChan() <-chan watch.Event {
+		graceful := func() bool {
+			c.stateMutex.Lock()
+			defer c.stateMutex.Unlock()
+			return c.wasBookmarkAfterResourceVersionReceivedLocked() && !c.wasBookmarkAfterResourceVersionSent()
```
> we are graceful when we have received an RV but i hasn't been sent, so we cannot simply check graceful := (c.state == cacheWatcherBookmarkReceived)

I don't understand. The semantic of cacheWatcherBookmarkReceived is exactly that:
- we already received a bookmark (with that RV)
- but we didn't send that yet

p0lyn0mial replied:
`graceful` indicates whether we should drain the input chan or not. 

we are draining the input chan only when
1.  we have received a bookmark with an RV >= the expected one
2. the bookmark hasn't been sent to the client

The semantic of `cacheWatcherBookmarkReceived` is `we already received a bookmark (with that RV)`, given the above we also need to check `2`

wojtek-t replied:
> The semantic of cacheWatcherBookmarkReceived is we already received a bookmark (with that RV), given the above we also need to check 2

No - as soon as we send the bookmark we call `markBookmarkAfterResourceVersionSentLocked` which is changing the state. So we actually are in this state exactly during that period.
This is the whole reason why I wanted state-machine - it's just very explicit.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -496,6 +509,32 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
+			// read the current RV under the lock
+			bookmarkAfterResourceVersion = c.watchCache.resourceVersion
+		default:
+			bookmarkAfterResourceVersion = watchRV
```
alternatively, we could check if the server has an RV >= the desired one and stop if it doesn't. In that case, we would have to close the current connection with the client so it could hit a different replica.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -496,6 +509,32 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
+	// Note that we cannot do it under Cacher lock, to avoid a deadlock, since the
+	// underlying watchCache is calling processEvent under its lock.
+	c.watchCache.RLock()
+	defer c.watchCache.RUnlock()
```
Why do we need to move the lock here?

IIUC, the only point where we really need that is when `watchRV==0` (to the the `watchCache.resourceVersion`).

If that's correct, then I would like to do that differently. There was a reason why the locking is happening relatively late - basically, to avoid memory allocations under lock (which historically was visibly increasing lock contention).
So what I would do basically is:
- revert the locking to where it was
- stop passing `bookmarkAfterResourceVersion` parameter as a newCacheWatcher parameter
- instead, set the bookmarkAfterResourceVersion field explicitly in the same place where its forget field is set

wojtek-t replied:
Modulo the fact that we should never be calling etcd under lock - that will be  disaster.

So what we need is basically something like that:
```
func (c *Cacher) getBookmarkAfterResourceVersionFunc(opts) func() uint64 {
  if !opts.SendInitialEvents && utilfeature... {
    return func() uint64 { return 0 }
  }
  switch {
  case len(opts.ResourceVersion) == 0:
    rv, err := c.getCurrentResourceVersionFromStorage(ctx)
    if err != nil {...}
    return func() uint64 { return rv }
  case watchRV == 0:
    // Here we assume that watchCache locked is already held.
    return func() uint64 { return c.watchCache.resourceVersion }
  default:
    return func() uint64 { return watchRV }
  }
}
```

p0lyn0mial replied:
yes, I moved the lock to get `watchCache.resourceVersion` for `watchRV==0`

> There was a reason why the locking is happening relatively late - basically, to avoid memory allocations under lock

are you referring to calling `newCacheWatcher` ? I think this is the only thing that has changed after moving the lock.

> Modulo the fact that we should never be calling etcd under lock - that will be disaster.

Could you please explain what it would be a disaster?

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -505,15 +544,9 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
-	// underlying watchCache is calling processEvent under its lock.
-	c.watchCache.RLock()
-	defer c.watchCache.RUnlock()
+	// Get initial events under the lock
```
Passing `watchRV` doesn't work now - you need to explicitly set watchRV=0 is SendInitialEvents is true.

p0lyn0mial replied:
not sure if I understand this comment.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -147,6 +148,10 @@ func (i *indexedWatchers) deleteWatcher(number int, value string, supported bool
 }
 
 func (i *indexedWatchers) terminateAll(objectType reflect.Type, done func(*cacheWatcher)) {
+	// note that we don't have to call setDrainInputBuffer method on the watchers
```
food for thought: for the sake of limiting size of this PR, maybe we can move support for `drainingWatchers` (without the logic deciding when to use it it ) to a separate PR and just merge it...

wojtek-t replied:
I went ahead and did that in: https://github.com/kubernetes/kubernetes/pull/111460

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:
```diff
@@ -66,6 +66,21 @@ type ListOptions struct {
 	// it has expired.
 	Continue string
+
+	// SendInitialEvents when set together with Watch option, begins the watch stream with synthetic "Added" events of all resources.
```
nit: can you please try keeping the lines a bit shorter (not longer then the longest lines in this file)?

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:
```diff
@@ -66,6 +66,21 @@ type ListOptions struct {
+	//
+	// - sendInitialEvent and RV>0 - provides the initial events and when we reach the desired RV,
+	//   the synthetic bookmark event. Note that the desired RV must be >= the current revision the server knows about.
+	//   If the desired RV is older than the current revision the server is aware of, the TooLargeResourceVersionError is returned.
```
That's not true - if the desired RV is older, then we simply return the more fresh state (which is fine according to "not older then" semantic. 
The TooLargeResourceVersionError is returned if the desired RV is actually more recent than anything server is aware of.

p0lyn0mial replied:
I just copied your previous comment https://github.com/kubernetes/kubernetes/pull/110960#discussion_r930792587

I have just checked and I think that we actually miss `TooLargeResourceVersionError` case. I think we should change the `getAllEventsSinceLocked` method with:
```
if resourceVersion > w.resourceVersion {
		return nil, storage.NewTooLargeResourceVersionError(resourceVersion, w.resourceVersion, resourceVersionTooHighRetrySeconds)
	}
```

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:
```diff
@@ -66,6 +66,21 @@ type ListOptions struct {
+	// This option is meant to be used with the resourceVersion parameter with the following semantics:
+	//
+	// - sendInitialEvent and RV>0 - provides the initial events and when we reach the desired RV,
+	//   the synthetic bookmark event. Note that the desired RV must be >= the current revision the server knows about.
```
how does the client distinguish between the bookmark indicating "initial events are done" and the bookmark indicating "server is shutting down"?

wojtek-t replied:
That's a good question.
The answer is that (at least by design) we will just not send the bookmark indicating "server shutting down" if we didn't already sync. I'm fairly sure the code works fine now, but we should add a test for that.

FWIW - we still don't really send bookmarks on shutdown though - we send them on "watch shutdown" only. But the question remains valid for that usecase (and the answer remains too).

p0lyn0mial replied:
in that special state, we won't send a bookmark that is < than the desired one even if the watcher is about to shut down.  I'll add a unit test to cover this case.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:
```diff
@@ -66,6 +66,21 @@ type ListOptions struct {
 	// it has expired.
 	Continue string
+
+	// SendInitialEvents when set together with Watch option, begins the watch stream with synthetic "Added" events of all resources.
```
The semantics described here make sense to me.  I think we should add logic that rejects this query param now, similar to what we did with dryrun.  That's a good start even if this doesn't make 1.25 and it gives us a clean PR to backport to supported versions to ease the transition.

wojtek-t replied:
That makes a lot of sense to me - agree with that.

p0lyn0mial replied:
moved to https://github.com/kubernetes/kubernetes/pull/111530

p0lyn0mial commented:
/remove-lifecycle stale

dims commented:
If you still need this PR then please rebase, if not, please close the PR

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1670,3 +1691,515 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
+					[]watch.Event{
+						{Type: watch.Added, Object: makePod(101)},
+						{Type: watch.Added, Object: makePod(102)},
+						{Type: watch.Bookmark, Object: &example.Pod{
```
atm this scenario fails because we don't send a bookmark event immediately, we could change a watcher (processInterval method) or change the scenario.

wojtek-t replied:
hmm - but verifyEvent is waiting couple seconds so it doesn't have to be immediate.

With realClock(), we should actually send it within a second.
So I don't understand why it is failing.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1104,6 +1121,80 @@ func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
+	}
+}
+
+func (c *Cacher) getWatchResourceVersion(ctx context.Context, parsedWatchResourceVersion uint64, opts storage.ListOptions) (uint64, error) {
```
we need a better name for this method.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1670,3 +1691,515 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
+		t.Fatal("failed adding an even to the watcher")
+	}
+
+	// the watcher is immediately expired
```
please explain why (it's waiting for bookmark so it scheduled immediately)

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1670,3 +1691,515 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
+				require.NoError(t, err, "failed to add a pod: %v")
+				// before starting a new watch set a storage RV to some future value
+				setStorageListMetaRV("102")
+				w, err := cacher.Watch(context.Background(), "pods/ns", storage.ListOptions{Predicate: func() storage.SelectionPredicate { p := storage.Everything; p.AllowWatchBookmarks = true; return p }(), SendInitialEvents: func() *bool { b := true; return &b }()})
```
Can you avoid those long line? Also let's do instead:
```
opts := storage.ListOptions{
  Labels: labels.Everything(),
  Fields: fields.Everything(),
  AllowWatchBookmarks: true.
  SendInitialEvents: &trueValue,
}
w, err := cacher.Watch(...)
```

[and define trueValue and falseValue above]

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1670,3 +1691,515 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
+				w, err := cacher.Watch(context.Background(), "pods/ns", storage.ListOptions{ResourceVersion: "101", Predicate: func() storage.SelectionPredicate { p := storage.Everything; p.AllowWatchBookmarks = true; return p }(), SendInitialEvents: func() *bool { b := true; return &b }()})
+				require.NoError(t, err, "failed to create watch: %v")
+				defer w.Stop()
+				// make sure we only get events form RV=101 and the bookmark
```
nit: from

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1670,3 +1691,515 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
+				verifyNoEvents(t, w)
+			},
+		},
+	}
```
I would really really like to avoid huge code-duplication here. These tests are almost the same, so I think we should switch to sth:

```
scenarios := []struct {
  name string
  eventsFirst     []watch.Event
  eventsSecond []watch.Event     // better names needed
}{
  {
     name: "allowWatchBookmarks=true, sendInitialEvents=true, RV=unset",
     events: ...
  }
  ...
}


for _, scenario := range scenarios {
  t.Run(....
    // roughly what you have now

   err := cacher.watchCache.Add(makePod(101))
   require.NoError(t, err, "failed to add a pod: %v")
   err = cacher.watchCache.Add(makePod(102))
   require.NoError(t, err, "failed to add a pod: %v")

   // set storage's RV to some future value
   storageListMetaResourceVersion = "105"
   opts := storage.ListOptions{ ....}
   w, err := cacher.Watch(context.Background(), "pods/ns", opts)
   require.NoError(t, err, "failed to create watch: %v")
   defer w.Stop()

   verifyEvents(...)
   verifyNoEvents(...)

   // add a pod that is greater than the storage's RV when the watch was started
  err = cacher.watchCache.Add(makePod(110))
  require.NoError(t, err, "failed to add a pod: %v")
  verifyEvents(...)
   verifyNoEvents()
```

It will require minor changes to some usecases, but it will be waaay easier to follow and maintain.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1670,3 +1691,515 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
+}
+
+// sendInitialEvents=true, RV=""
+func TestCacheWatchSemanticsScenario1(t *testing.T) {
```
Is this test really needed - I think this is now tested in the test above.

p0lyn0mial replied:
yes, cruft, sorry, will remove.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1670,3 +1691,515 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
 		t.Fatalf("expected forget() to be called three times, because processInterval should call Stop(): %v", err)
 	}
 }
+
```
The tests are good, but I would like to see one more test that will be exercising `getCurrentResourceVersionFromStorage` functionality.

wojtek-t reviewed (commented):
I added a lot of comments, but I think we're actually relatively close.

p0lyn0mial committed:
cacher: Add support for consistent streaming

design details https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3157-watch-list#design-details

p0lyn0mial committed:
cacher_watcher: Add support for consistent streaming

design details https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3157-watch-list#design-details

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1104,6 +1121,80 @@ func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
+	pred := storage.SelectionPredicate{
+		Label: labels.Everything(),
+		Field: fields.Everything(),
+		Limit: 1, // just in case we actually hit something
```
I think we should by fine. 

`Recursive: true` seems to be setting a prefix for the key (`WithPrefix`).
By default `Get/List` requests are linearizable and users have to explicitly ask for stale reads with `WithSerializable`.

```
// WithSerializable makes 'Get' request serializable. By default,
// it's linearizable. Serializable requests are better for lower latency
// requirement.
func WithSerializable() OpOption {
	return func(op *Op) { op.serializable = true }
}
```

wojtek-t replied:
OK - let's just have a test for it that has a real etcd underneath propagated with some data and let's ensure that the object that is returned is of lower RV (but still we get the right higher RV as part of list meta).

wojtek-t replied:
The test is there - resolving.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1275,28 +1394,60 @@ func (c *cacheWatcher) stopLocked() {
+		graceful := func() bool {
+			c.stateMutex.Lock()
+			defer c.stateMutex.Unlock()
+			return c.wasBookmarkAfterResourceVersionReceivedLocked() && !c.wasBookmarkAfterResourceVersionSentLocked()
```
I'm not sure if I understand. We can terminate the watch (`false`) when both `a` and `b` from the comment above applies.

wojtek-t replied:
wasBookmarkAfterResourceVersionReceived = `c.state != cacheWatcherWaitingForBookmark`
~wasBookmarkAfterResourceVersionSent = `! c.state == cacheWatcherBookmarkSent`

So effectively `&&` of those two conditions is:
```
c.state != cacheWatcherWaitingForBookmark && c.state != cacheWatcherBookmarkSent
```
which is equal to
```
c.state == cacheWatcherBookmarkReceived
```

[Sorry - I think there was a typo in my original comment.]

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1376,7 +1602,21 @@ func updateResourceVersion(object runtime.Object, versioner storage.Versioner, r
 	if event.Type == watch.Bookmark {
-		return &watch.Event{Type: watch.Bookmark, Object: event.Object.DeepCopyObject()}
+		e := &watch.Event{Type: watch.Bookmark, Object: event.Object.DeepCopyObject()}
+		if !c.wasBookmarkAfterResourceVersionSent() {
```
Discussed offline.

This is only happening for bookmarks (which are rare), this is per-watch lock, so contention isn't a problem so altogether that isn't needed.

Let's leave mutex as is - it's easier to reason and less error-prone.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1670,3 +1691,515 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
+		defer lock.RUnlock()
+		return count == 1, nil
+	}); err != nil {
+		t.Fatalf("expected forget() to be called once, because processInterval should call Stop(): %v", err)
```
yes, you are right, it should be called twice, I'll update the check.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1670,3 +1691,515 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
+		defer lock.RUnlock()
+		return count == 1, nil
+	}); err != nil {
+		t.Fatalf("expected forget() to be called once, because processInterval should call Stop(): %v", err)
```
no, here it is different, here we are in the draining mode, compare with https://go.dev/play/p/W77KHWObERF

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1670,3 +1691,515 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
+	var lock sync.RWMutex
+	var w *cacheWatcher
+	watchInitializationSignal := utilflowcontrol.NewInitializationSignal()
+	ctx := utilflowcontrol.WithInitializationSignal(context.Background(), watchInitializationSignal)
```
we need to make sure that the goroutine has been scheduled otherwise we might fail on adding the 3rd event

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1104,6 +1121,80 @@ func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
+}
+
+func (c *Cacher) getWatchResourceVersion(ctx context.Context, parsedWatchResourceVersion uint64, opts storage.ListOptions) (uint64, error) {
+	if opts.SendInitialEvents == nil || *opts.SendInitialEvents == true {
```
discussed offline.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1104,6 +1121,80 @@ func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
+	return uint64(currentResourceVersion), nil
+}
+
+func (c *Cacher) getBookmarkAfterResourceVersionFunc(ctx context.Context, parsedResourceVersion uint64, opts storage.ListOptions) (func() uint64, error) {
```
if that is okay with you I prefer a method, it aligns with `getStartResourceVersionForWatchLocked`

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_test.go:
```diff
@@ -0,0 +1,104 @@
+	"k8s.io/apiserver/pkg/storage/value/encrypt/identity"
+)
+
+func TestGetCurrentResourceVersionFromStorage(t *testing.T) {
```
This is good test - but I would just ask for moving it to cacher_whitebox_test.go instead of creating a new file here (you're effectively testing it in whitebox way looking into cacher internals).

p0lyn0mial replied:
okay, I don't know why but I though the other tests are more like black box, will move, thx.

wojtek-t reviewed (commented):
This is great - I still have few comments, but those are much smaller at this point.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cache_watcher_test.go:
```diff
@@ -412,3 +415,177 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
+	if !w.stopped {
+		t.Fatal("expected the watcher to be stopped but it wasn't")
+	}
+	verifyEvents(t, w, []watch.Event{{Type: watch.Added, Object: makePod(1)}, {Type: watch.Added, Object: makePod(2)}}, false)
```
because initial events are stored in a map and listing from a map give random order.

wojtek-t replied:
in general - yes.
But here you're explicitly using `intervalFromEvents` which isn't using a map, no?

p0lyn0mial replied:
ah, right, I thought your comment applies a scenario from `TestWatchSemantics` test.
here, indeed we populate the watcher from a slice so the order will always be the same.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cache_watcher_test.go:
```diff
@@ -412,3 +415,177 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
+	var lock sync.RWMutex
+	var w *cacheWatcher
+	watchInitializationSignal := utilflowcontrol.NewInitializationSignal()
+	ctx := utilflowcontrol.WithInitializationSignal(context.Background(), watchInitializationSignal)
```
we need to make sure that the goroutine has been scheduled otherwise we might fail on adding the 3rd event (nothing will start processing the input chan)

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -542,6 +553,10 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 	// underlying watchCache is calling processEvent under its lock.
 	c.watchCache.RLock()
 	defer c.watchCache.RUnlock()
+	watchRV, err = c.getStartResourceVersionForWatchLocked(ctx, watchRV, opts)
```
ok, good catch, we need something similar to `getBookmarkAfterResourceVersionLockedFunc`

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1367,3 +1416,200 @@ func TestCacheIntervalInvalidationStopsWatch(t *testing.T) {
+		expectedEventsAfterEstablishingWatch []watch.Event
+	}{
+		{
+			name:                       "allowWatchBookmarks=true, sendInitialEvents=true, RV=unset, storageRV=102",
```
sure, I can do that but it looks so sleek with the `name` field :)

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1367,3 +1419,276 @@ func TestCacheIntervalInvalidationStopsWatch(t *testing.T) {
+	scenarios := []struct {
+		name                   string
+		allowWatchBookmarks    bool
+		sendInitialEvents      string
```
Instead of making it string, let's make it *bool

[you define
trueVal := true
falseVal := false
first and then can use pointers to those later]

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1157,6 +1177,85 @@ func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
+}
+
+// getBookmarkAfterResourceVersionLockedFunc returns a ResourceVersion after which the bookmark event will be delivered.
+// This method must be called under the watchCache lock.
```
this isn't true - the function that is returned must be called under watchCache lock.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1157,6 +1177,85 @@ func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
+//   - start at Most Recent (return an RV from etcd)
+//   - start at Any (return the current watchCache's RV)
+//
+// This method must be called under the watchCache lock.
```
Same here

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -498,11 +503,11 @@ type namespacedName struct {
 	pred := opts.Predicate
-	// If the resourceVersion is unset, ensure that the rv
-	// from which the watch is being served, is the latest
+	// if the watch-list feature wasn't set and the resourceVersion is unset
```
Just to be really on the safe side - let's just add:
```
if utilfeature.DefaultFeatureGate.Enabled(features.WatchList) && opts.SendinitialEvents != nil {
  opts.SendInitialEvents = nil
}
```

[I know that technically it should never happen due to our validation, but let's just be super-safe here.]

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -863,6 +864,13 @@ func growSlice(v reflect.Value, maxCapacity int, sizes ...int) {
 
 // Watch implements storage.Interface.Watch.
 func (s *store) Watch(ctx context.Context, key string, opts storage.ListOptions) (watch.Interface, error) {
+	if opts.SendInitialEvents != nil {
```
I don't fully agree - it requires client-changes.

But as discussed offline - I'm fine with it for Alpha, so let's talk about it after this PR is merged.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -1082,28 +1094,68 @@ func TestDispatchEventWillNotBeBlockedByTimedOutWatcher(t *testing.T) {
+	select {
+	case e := <-w.ResultChan():
+		t.Errorf("Unexpected: %#v event received, expected no events", e)
+	default:
```
Let's switch to actually wait for a bit of time (like 1s).
So sth like
```
case <-time.After(time.Second):
```

wojtek-t reviewed (commented):
I'm sorry - just 2 more nits and with that I'm fine with this being merged.

Once fixed, I will kick in merging the API PR and ask you to rebase this one and have this merged.

wojtek-t commented:
OK - this LGTM - thanks a lot for bearing with me!

I'm tagging this and kicking the API to merge, so that we can rebase this one and merge.

/lgtm
/approve

wojtek-t commented:
/lgtm

Thanks!
